### PR TITLE
Build a Debian package around the static binary

### PR DESCRIPTION
**Pull request by tobim**
It is now possible to build a `package` target to create Debian package, for example `vast_2.3.0-1_amd64.deb`.

### Todo:

- [x] Change Nix to do a package only build to avoid creating the binaries twice
- [x] Upload the package to GCS & Extend installation instructions.

### :memo: Reviewer Checklist

Review this pull request by ensuring the following items:

- [x] All user-facing changes have changelog entries
- [x] User-facing changes are reflected on [vast.io](https://github.com/tenzir/vast/tree/master/web)

### CODE REVIEW

**Review by mavam — COMMENTED** [review]
Maybe we can move the `debian/` dir to `scripts/debian/`?

File: debian/postinst
Comment (dominiklohmann) [inline debian/postinst:28]: If the postinstall script references our systemd service then we should actually install it like we do for rc.d and launchd.

File: cmake/Package.cmake
Comment (dominiklohmann) [inline cmake/Package.cmake:3]: Shouldn't this respect the CMAKE_INSTALL_PREFIX?

File: nix/vast/default.nix
Comment (tobim) [reply nix/vast/default.nix:143]: 1. Afaik resetting these variables on an existing build tree has no effect.
2. Nix build anything incrementally.

File: nix/static-binary.sh
Comment (tobim) [reply nix/static-binary.sh:108]: I can change that.

File: cmake/Package.cmake
Comment (tobim) [reply cmake/Package.cmake:10]: No, `/var` is always `/var`. At least that is how I understand the FHS.

File: cmake/Package.cmake
Comment (tobim) [reply cmake/Package.cmake:3]: Yeah I think we can just remove that.

File: CMakeLists.txt
Comment (tobim) [reply CMakeLists.txt:748]: The current custom tar.gz thing is broken because the nix build doesn't do a regular install anymore. The way to go is to use CPack to create that as well. Once we have component based installation the output should be equivalent.

File: CMakeLists.txt
Comment (tobim) [reply CMakeLists.txt:16]: Ah, I didn't know that is used somewhere. I just thought `vast.io` is more appropriate.

File: CMakeLists.txt
Comment (tobim) [reply CMakeLists.txt:750]: I changed things so we now get a `.tar.zst` from CPack.

File: nix/vast/default.nix
Comment (tobim) [reply nix/vast/default.nix]: The version you saw was already out of date when you commented. These `-U`s are now done as part of the configuration step. 
The function `stdenv.mkDerivation` doesn't have an explicit packaging phase, so we create a *package* variant instead now.

File: scripts/dbsize.sh
Comment (dominiklohmann) [inline scripts/dbsize.sh:3]: This requires GNU coreutils' version of `df`, so in order to at least also make this work on macOS (which I would much appreciate for testing actually), please change it like so:

```bash
PATH="$(brew --prefix coreutils)/libexec/gnubin:${PATH}" df --output=pcent . | tr -dc '0-9'
```

(Or make it a bit smarter to not always modify $PATH, but instead to only do so if `brew --prefix coreutils` doesn't fail.)

File: CMakeLists.txt
Comment (dominiklohmann) [inline CMakeLists.txt:727]: We should probably rename that script to `vast-du-pcent.sh` or so, and add an explanatory comment to inside the script given that we now install it.

File: CMakeLists.txt
Comment (dominiklohmann) [inline CMakeLists.txt:750]: Can you rename the file from `Package.cmake` to `VASTPackage.cmake`? This has become common practice to be able to disambiguate which project a CMake include came from.

File: cmake/Package.cmake
Comment (dominiklohmann) [inline cmake/Package.cmake:21]: Should this be removed before merging?

File: cmake/Package.cmake
Comment (dominiklohmann) [inline cmake/Package.cmake:3]: We might wanna put Tenzir GmbH here, not sure exactly.

File: CMakeLists.txt
Comment (dominiklohmann) [inline CMakeLists.txt:732]:
```suggestion
          DESTINATION "${CMAKE_INSTALL_LIBEXECDIR}")
```

File: cmake/Package.cmake
Comment (dominiklohmann) [inline cmake/Package.cmake:11]:
```suggestion
set(CPACK_GENERATOR "TZST")

if (${CMAKE_SYSTEM_NAME} MATCHES "Linux")
  list(APPEND CPACK_GENERATOR "DEB")
endif ()
```

File: cmake/Package.cmake
Comment (dominiklohmann) [inline cmake/Package.cmake:14]:
```suggestion
set(CPACK_DEBIAN_COMPRESSION_TYPE "gzip")
set(CPACK_DEBIAN_FILE_NAME "DEB-DEFAULT")
set(CPACK_DEBIAN_PACKAGE_NAME "vast")
```

For consistency with the rest of our CMake.

**Review by dominiklohmann — COMMENTED** [review]
This is looking great. I had some minor comments inline, and modulo manual testing this is ready to go from my side. Obviously still needs the Runtime/Development component split to be perfect, but I'll do that in a separate PR once this is merged to avoid the conflicts.

Thanks so much for working on this! 🎉

File: cmake/Package.cmake
Comment (dominiklohmann) [reply cmake/Package.cmake:9]: I stand corrected. Thanks for double-checking.

File: nix/vast/default.nix
Comment (dominiklohmann) [reply nix/vast/default.nix]: Yeah this is looking much better now. Thanks!

File: CMakeLists.txt
Comment (dominiklohmann) [reply CMakeLists.txt:750]: `git grep \.tar\.gz` still yields quite a few links that will be broken after the next release with this merged. Are you sure we don't just wanna stick with the more widely available `.tar.gz`?

Comment (tobim) [reply CMakeLists.txt:750]: Reverted to tar.gz for link stability. 
Although `zst` is commonly accepted in most places nowadays.

File: cmake/Package.cmake
Comment (tobim) [reply cmake/Package.cmake:21]: I would keep it. The extra output isn't overly verbose:
```
CPackDeb:Debug: CPACK_TOPLEVEL_DIRECTORY          = '/home/tobim/t/vast/deb/build/gcc-static/redeb/_CPack_Packages/Linux/DEB'
CPackDeb:Debug: CPACK_TOPLEVEL_TAG                = 'Linux'
CPackDeb:Debug: CPACK_TEMPORARY_DIRECTORY         = '/home/tobim/t/vast/deb/build/gcc-static/redeb/_CPack_Packages/Linux/DEB/VAST-2.3.0-Linux'
CPackDeb:Debug: CPACK_OUTPUT_FILE_NAME            = 'vast_2.3.0-1_amd64.deb'
CPackDeb:Debug: CPACK_OUTPUT_FILE_PATH            = '/home/tobim/t/vast/deb/build/gcc-static/redeb/vast_2.3.0-1_amd64.deb'
CPackDeb:Debug: CPACK_PACKAGE_FILE_NAME           = 'VAST-2.3.0-Linux'
CPackDeb:Debug: CPACK_PACKAGE_INSTALL_DIRECTORY   = 'VAST 2.3.0'
CPackDeb:Debug: CPACK_TEMPORARY_PACKAGE_FILE_NAME = '/home/tobim/t/vast/deb/build/gcc-static/redeb/_CPack_Packages/Linux/DEB/vast_2.3.0-1_amd64.deb'
CPackDeb:Debug: CPACK_DEBIAN_PACKAGE_CONTROL_STRICT_PERMISSION = ''
CPackDeb:Debug: CPACK_DEBIAN_PACKAGE_SOURCE       = ''
```

File: cmake/Package.cmake
Comment (tobim) [reply cmake/Package.cmake:3]: "Tenzir" should be enough.

File: cmake/Package.cmake
Comment (tobim) [reply cmake/Package.cmake:11]: The Generator selection is now done from the command line.

File: CMakeLists.txt
Comment (tobim) [reply CMakeLists.txt]: I renamed it to `vast-df-percent.sh` and modified it for portability.

File: scripts/debian/postinst
Comment (dominiklohmann) [inline scripts/debian/postinst:27]: Is there documentation for this that you can add as a comment?

**Comment by tobim** [comment]
This is now already complete, I already tested `dpkg -i`, `apt remove`, and `apt purge`.

The last things I want to add are
- [ ] the `-rc` suffix is missing from the deb filename.
- [ ] The deb package from the debian ci build should be uploaded as artifact
- [ ] The docs need a slight adaptation

I think I can check these off this evening.

File: CMakeLists.txt
Comment (dominiklohmann) [inline CMakeLists.txt:102]: Unnecessary newline that you introduced here.

File: nix/overlay.nix
Comment (dominiklohmann) [inline nix/overlay.nix:100]: Both the systemd and the debian directory moved.

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:302]: The semicolon requires you to add quotes, otherwise bash does the wrong thing here.

File: vast/CMakeLists.txt
Comment (dominiklohmann) [inline vast/CMakeLists.txt:69]: Now that we have this logic we should remove the `systemd` directory from the repository root and move the remaining contents over to `vast/services/systemd` as well.

This also requires changing one line in the Dockerfile in the repository root, which references `systemd/vast.yaml`.

Comment (dominiklohmann) [inline vast/CMakeLists.txt:69]: This needs to be covered in the documentation under `web/docs/setup-vast/install/linux.md`, which now links to the wrong files and contains outdated information.

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:975]: We should update the docs on `web/docs/setup-vast/download.md` to reference the Deb in addition to the tar.gz.

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:302]: The additional Deb should also be uploaded for to GitHub as an artifact and for releases for the `build-debian`  job.

**Review by dominiklohmann — COMMENTED** [review]
Some final minor remarks. This works really well in practice, so once this round is through I'd want to just have one additional docs review.

This also needs changelog entries still, I think:
- With the next release we'll start providing .deb packages.
- The newly bundled `vast-df-percent.sh` script needs to be explained.
- VAST now installs a systemd unit on Linux by default.

File: vast/CMakeLists.txt
Comment (tobim) [reply vast/CMakeLists.txt]: Why are we copying the service file into the docker container?

**Comment by tobim** [comment]
I didn't get around to updating the docs yet, and the CI copy changes are probably not correct yet either.

**Comment by mavam** [comment]
Given the complexity of this PR and new developer tooling, we probably should add some details on how to build the package in https://vast.io/docs/develop-vast/contributing/build-environment.

File: vast/CMakeLists.txt
Comment (dominiklohmann) [reply vast/CMakeLists.txt]: We are not, we're just re-using the same `vast.yaml` file that we reference by default in the service file.

**Comment by lava** [comment]
Can we reduce the scope of this PR by not actually providing *public* debian packages yet but only producing them as an additional CI artifact that can be used in our downstream workflows? Having them available for users is a nice extra, but it feels like spending all the time on getting the documentation right is delaying the important part.

Speaking of artifacts, the current version only produces a single file called `artifact` which doesn't look right, although it already has the correct files:
```
$ unzip artifact.zip 
Archive:  artifact.zip
  inflating: vast-v2.3.0-rc2-29-ge0e580cf1a-Linux.tar.gz  
  inflating: vast_v2.3.0-rc2-29-ge0e580cf1a_amd64.deb  
```

**Comment by dominiklohmann** [comment]
> Can we reduce the scope of this PR by not actually providing _public_ debian packages yet but only producing them as an additional CI artifact that can be used in our downstream workflows? Having them available for users is a nice extra, but it feels like spending all the time on getting the documentation right is delaying the important part.

So far we haven't spent any time on that, and I personally think that getting that right should be part of this PR so we don't ship something that's undocumented.

Whether we only attach to releases or only produce for downstream workflows is mostly irrelevant, as the former is already done and implemented.



> Speaking of artifacts, the current version only produces a single file called `artifact` which doesn't look right, although it already has the correct files:
> 
> ```
> $ unzip artifact.zip 
> Archive:  artifact.zip
>   inflating: vast-v2.3.0-rc2-29-ge0e580cf1a-Linux.tar.gz  
>   inflating: vast_v2.3.0-rc2-29-ge0e580cf1a_amd64.deb  
> ```

This requires calling the upload-artifact action twice instead of once with both files iirc, should be easy to find. The per-plugin CI will fail with this broken also, because it attempts to download the tar.gz artifact from the preceding runs.

File: vast/CMakeLists.txt
Comment (dominiklohmann) [reply vast/CMakeLists.txt]: We're doing this in a follow-up PR.

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [reply .github/workflows/vast.yaml]: We've decided to instead explicitly disable DEB + non-static builds.

**Review by dominiklohmann — APPROVED** [review]
Changelog entry still missing, other than that LGTM.

File: web/docs/setup-vast/install/linux.md
Comment (dominiklohmann) [inline web/docs/setup-vast/install/linux.md:82]:
```suggestion
The Debian package automatically creates a `vast` system user and installs
```

File: web/docs/setup-vast/install/linux.md
Comment (dominiklohmann) [inline web/docs/setup-vast/install/linux.md:76]:
```suggestion
and Debian-derived distributions. After downloading, install VAST using `dpkg`:
```

File: changelog/unreleased/features/2513--debian-package.md
Comment (dominiklohmann) [inline changelog/unreleased/features/2513--debian-package.md:7]: Please rewrite using active voice, and link to the GitHub releases page rather than our public GCS bucket.

File: .github/workflows/vast.yaml
Comment (dominiklohmann) [inline .github/workflows/vast.yaml:310]: This triggers a warning in CI, so I think we should add a comment on why we're adding this.

**Review by dominiklohmann — COMMENTED** [review]
Some additional minor things I've come across when working on #2544.

**Pull request merged**
